pkg/istiodcert/worker.go

```diff
@@ -0,0 +1,281 @@
+	return true
+}
+
+func (dicp *DynamicIstiodCertProvisioner) handleNewIssuer(issuerRef *cmmeta.ObjectReference) {
+	dicp.issuerRefMutex.Lock()
+	defer dicp.issuerRefMutex.Unlock()
+
+	if issuerRef == nil && dicp.initialIssuerRef != nil {
+		// don't blank out the issuer if there's an initial ref; use that instead
+		dicp.issuerRef = dicp.initialIssuerRef
+		return
+	}
+
+	dicp.issuerRef = issuerRef
+
+	dicp.log.Info("triggering reconciliation of istiod cert after issuer change", "cert_name", dicp.opts.CertificateName, "cert_namespace", dicp.opts.CertificateNamespace)
+
+	dicp.reconcileChan <- event.GenericEvent{
+		Object: &cmapi.Certificate{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      dicp.opts.CertificateName,
+				Namespace: dicp.opts.CertificateNamespace,
+			},
+		},
+	}
+}
+
+// AddControllersToManager adds controllers to the given manager which:
+// 1. Handle provisioning and updating the dynamic istiod cert
+// 2. Handle listening for updates to the active issuer ref and re-issuing
+func (dicp *DynamicIstiodCertProvisioner) AddControllersToManager(mgr manager.Manager) error {
+	b := ctrl.NewControllerManagedBy(mgr)
+
+	b.For(
+		new(cmapi.Certificate), builder.WithPredicates(predicate.NewPredicateFuncs(func(obj client.Object) bool {
+			// Only process one specific cert which was requested
+			return obj.GetName() == dicp.opts.CertificateName && obj.GetNamespace() == dicp.opts.CertificateNamespace
+		})))
+
+	// when the issuer changes, trigger a re-reconciliation
+	b.WatchesRawSource(source.Channel(dicp.reconcileChan, handler.EnqueueRequestsFromMapFunc(
+		func(context.Context, client.Object) []reconcile.Request {
+			return []reconcile.Request{ctrl.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      dicp.opts.CertificateName,
+					Namespace: dicp.opts.CertificateNamespace,
+				},
+			}}
+		})))
+
+	err := b.Complete(dicp)
+	if err != nil {
+		return err
+	}
+
+	err = mgr.Add(dicp)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (dicp *DynamicIstiodCertProvisioner) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	dicp.issuerRefMutex.Lock()
+	defer dicp.issuerRefMutex.Unlock()
+
+	if dicp.issuerRef == nil {
+		dicp.log.Info("exiting reconcile of dynamic istiod early; no issuerRef is set")
+		return ctrl.Result{}, nil
+	}
+
+	dicp.log.Info("reconciling dynamic istiod cert", "issuer-name", dicp.issuerRef.Name, "issuer-kind", dicp.issuerRef.Kind, "issuer-group", dicp.issuerRef.Group)
+
+	spiffeID := fmt.Sprintf("spiffe://%s/ns/%s/sa/istiod-service-account", dicp.trustDomain, req.Namespace)
+
+	commonName, dnsNames := makeDNSNamesFromRevisions(req.Namespace, dicp.opts.IstioRevisions)
+
+	if len(dicp.opts.AdditionalDNSNames) > 0 {
+		dnsNames = append(dnsNames, dicp.opts.AdditionalDNSNames...)
+	}
+
+	privateKeyAlg := cmapi.RSAKeyAlgorithm
+	if dicp.opts.KeyAlgorithm == "ECDSA" {
+		privateKeyAlg = cmapi.ECDSAKeyAlgorithm
+	}
+
+	desiredSpec := cmapi.CertificateSpec{
```

Comment:
note for reviewers: The intent of this spec is to match what's in the static cert (https://github.com/cert-manager/istio-csr/blob/cb8edf6f76cb2a85c332676c5789ff1dca46a3b2/deploy/charts/istio-csr/templates/certificate.yaml)

I think this should be identical to that cert. It certainly seemed to be in testing.